client_it_test.go

```diff
@@ -92,6 +95,163 @@ func TestClientLifecycleEvents(t *testing.T) {
 	})
 }
 
+func TestClient_AddLifecycleListener(t *testing.T) {
+	receivedStates := []hz.LifecycleState{}
+	receivedStatesMu := &sync.RWMutex{}
+	it.Tester(t, func(t *testing.T, client *hz.Client) {
+		subscriptionID, err := client.AddLifecycleListener(func(event hz.LifecycleStateChanged) {
+			receivedStatesMu.Lock()
+			defer receivedStatesMu.Unlock()
+			// client is already connected
+			switch event.State {
+			case hz.LifecycleStateShuttingDown:
+				t.Logf("Received shutting down state")
+			case hz.LifecycleStateShutDown:
+				t.Logf("Received shut down state")
+			default:
+				t.Log("Received unknown state:", event.State)
+			}
+			receivedStates = append(receivedStates, event.State)
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		require.NotEqual(t, types.UUID{}, subscriptionID, "subscription UUID should not be empty")
+		if err = client.Shutdown(context.Background()); err != nil {
+			t.Fatal(err)
+		}
+		defer func() {
+			receivedStatesMu.Lock()
+			receivedStates = []hz.LifecycleState{}
+			receivedStatesMu.Unlock()
+		}()
+		targetStates := []hz.LifecycleState{
+			hz.LifecycleStateShuttingDown,
+			hz.LifecycleStateShutDown,
+		}
+		it.Eventually(t, func() bool {
+			receivedStatesMu.RLock()
+			defer receivedStatesMu.RUnlock()
+			return reflect.DeepEqual(targetStates, receivedStates)
+		})
+	})
+}
+
+func TestClient_RemoveLifecycleListener(t *testing.T) {
+	var lifecycleEventReceived int32 = 1
+	it.Tester(t, func(t *testing.T, client *hz.Client) {
+		subscriptionID, err := client.AddLifecycleListener(func(event hz.LifecycleStateChanged) {
+			// shutting down event is published before the actual shutdown is occurred
+			// client must not be shutdown before published that event
+			if event.State == hz.LifecycleStateShuttingDown {
+				atomic.SwapInt32(&lifecycleEventReceived, 0)
+			}
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		assert.NotEqual(t, types.UUID{}, subscriptionID, "subscription UUID should not be empty")
+		if err = client.RemoveLifecycleListener(subscriptionID); err != nil {
+			t.Fatal(err)
+		}
+		if err = client.Shutdown(context.Background()); err != nil {
+			t.Fatal(err)
+		}
+		it.Eventually(t, func() bool {
+			return !client.Running() && lifecycleEventReceived == 1
+		})
+	})
+}
+
+func TestClient_AddMembershipListener(t *testing.T) {
+	var (
+		wgAdded,
+		wgRemoved sync.WaitGroup
+		memberCount int = 3
+	)
+	wgAdded.Add(1)
+	wgRemoved.Add(1)
+	ctx := context.Background()
+	cls := it.StartNewClusterWithOptions("client-subscribed-add-membership-listener", 15701, memberCount)
```

Comment:
Could you defer shutdown the cluster? Same comment for similar code.

Reply:
Fixed: [88b6f60](https://github.com/hazelcast/hazelcast-go-client/pull/809/commits/88b6f60a915c3d0237ae5f1fb97bdbda52cedfbd)